### ISSUE TITLE
Change interface of priority functions for easier "framework-based" parallelization

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#24246]
I'm currently trying to improve the overall scheduling throughput. It seems we can achieve significant gains from different kind of parallelization.
However, I would like to avoid parallizing individual predicate/priority function and support parallelization at the framework level.

For predicate function, it is easy to parallelize computations of them between nodes, since they are independent.
However, it's not the case for priorities - the interface of priority function is that it computes priorities for all nodes at the same time.

It would be much easier to parallelize everything, when priority function would also take a node as an argument and would compute priority wrt to current priority function just for this particular node.
However, there are two problematic functions for now (maybe more in the future):
- SelectorSpreading
- NodeAffinity

Both of them have the same main problem - they are computing some partial scores for nodes independently, but then to compute the final score for the node, they rely on the maximum of those partial scores from all nodes.
### Proposal
1. I'm proposing to change the interface of priority functions to take a node for which they are supposed to compute priority.
2. We also extend the HostPriority type to include "priorityFunctionName" string
3. We will split priority functions into two phases:
4. per-node computations (which are generally cpu-intensive and would benefit from parallelization)
5. aggregation (which is generally much cheaper) and we can afford not parallelizing it

Thoughts?

@davidopp @kubernetes/sig-scheduling @kubernetes/sig-scalability

**Comment by gmarek** [comment]
IIUC the reason for priority functions to take list of nodes instead of single nodes was exactly what you mentioned - allowing priority function to be relative. Having said that, I think that 'map-reduce' way of doing it is probably a good optimization, as it should be expressive enough (and for all current priorities it certainly is).

I'm not sure if arguments of priority function are not part of the contract, so this change will be backward compatible.

**Comment by hongchaodeng** [comment]
Priority score is per node.
The dependencies could be modeled as DAG computation flow. More parallelization should be achievable.

/cc @xiang90

**Comment by gmarek** [comment]
The question is how much we need dependencies. Maybe single "map-reduce" will suffice. Can you give an example of sane priority function that requires full DAG?

**Comment by wojtek-t** [comment]
I don't think we should do a full dag - this is overengineering.

**Comment by davidopp** [comment]
@wojtek-t I haven't thought carefully about it, but that approach seems reasonable (assuming you think the speedup justifies the extra complexity). BTW CalculateInterPodAffinityPriority() in 
https://github.com/kubernetes/kubernetes/pull/22985/files
has the same issue (that PR is still in review).

**Comment by wojtek-t** [comment]
@davidopp - thanks

>
(assuming you think the speedup justifies the extra complexity).

I wouldn't call it just "extra complexity". 
It will definitely complicate the framework, but at the same time, I would say it will simplify the individual priority functions. If so, then I think it's a gain from productivity performance too, because we should very rarely touch framework itself, but we may want to add/modify priority/predicate functions much more often.

Anyway - I will try to create a POC PR next week and we can discuss further on it.

**Comment by hongchaodeng** [comment]
SGTM.
I like the fact that it will achieve some parallelization while simplifying individual implementation.

**Comment by davidopp** [comment]
Please don't merge any PR for this until #22985 has gone in. That PR is a beast, and I'd rather get that in first, then modify its priority function to fit the "map-reduce" model (I like Marek's terminology :-) )

**Comment by wojtek-t** [comment]
@davidopp - sure;
In fact on Friday I was doing some experiments and I found that we have some periods of time (in orders of 10-30ms), when scheduler is not doing any useful work. Since it's Go 1.6, it's shouldn't be "stop-theworld" GC, so I'm doing to investigate it first.
Anyway, my goal for now is to how long every single phase of scheduling takes (predicates, priorities, binding, choosing pod for scheduling, etc.) with size of cluster/number of existing pods as an argument and unerstand exactly why is that (I still don't understand some of its pieces deeply).

**Comment by davidopp** [comment]
This wasn't done for 1.3, moving to 1.4 milestone.

**Comment by dims** [comment]
Does this slip to 1.5?

**Comment by davidopp** [comment]
Looks like it will be in 1.5. See #31606.

**Comment by wojtek-t** [comment]
Yes - I would like to have it in 1.5.

**Comment by HardySimpson** [comment]
hi, all, I have a idea about the interface 


This is the interface now

```
func LeastRequestedPriority(pod *api.Pod, nodeNameToInfo map[string]*schedulercache.NodeInfo, nodes []*api.Node) (schedulerapi.HostPriorityList, error) 
```

if we turn the slice of node to a channel, then we can both achieve the goal of parallelization and simple implementation

```
func LeastRequestedPriority(pod *api.Pod, nodeNameToInfo map[string]*schedulercache.NodeInfo, nodes <-*api.Node) (schedulerapi.HostPriorityList, error)  {
	podResources := getNonZeroRequests(pod)
	list := make(schedulerapi.HostPriorityList, 0, 10)
        // The for each statement NOT change here, and we can put node for which they are supposed to compute priority.
	for _, node := range nodes {
		list = append(list, calculateUnusedPriority(pod, podResources, node, nodeNameToInfo[node.Name]))
	}
	return list, nil
}


```

**Comment by wojtek-t** [comment]
@HardySimpson - your proposal won't solve the problem of equivalence classes that we are facing there - see https://github.com/kubernetes/kubernetes/pull/31605#discussion_r76994422

The other counter argument for that is that channels introduce some overhead - I did some simple experiments few months ago around that (I also though about it), and it was visibly slower.

**Comment by HardySimpson** [comment]
@wojtek-t the interface is not only for solve equivalence classes, it can take any node from predicate function immediately, and calc it priority if possible.

what is the channel cost? to much CPU or memory? or cause lantency?

**Comment by wojtek-t** [comment]
>
@wojtek-t the interface is not only for solve equivalence classes, it can take any node from predicate function immediately, and calc it priority if possible.

As I said - I did this experiment few months ago and this is slower.

**Comment by gmarek** [comment]
MapReduce kind of computations are trivially shardable, so there's no point of using channels.

**Comment by dims** [comment]
This needs to be triaged as a release-blocker or not for 1.5 @wojtek-t

**Comment by wojtek-t** [comment]
This is not a release blocker for 1.5.

**Comment by k82cn** [comment]
>> @wojtek-t , one more input for this is to add an interface for those algorithm to include Name, ConditionChanged, Mandatory, e.g. for kubelet and scheduler, both of them should enforce mandatory predicates; for equal cache, we can ask algorithm whether the condition it care where changed.

> Sorry, I think I'm not really following. Can you clarify?

Sorry for the delay. I'm thinking the interface for Predicates :).

```go
type Predicate interface {
  Name() string
  // Check whether the condition, impact the predicate,  was changed; 
  // this's used by EqualCache, to avoid the dependency in factor.go to check condition
  // for each predicate.
  ConditionChanged(old, obj interface{}) bool
  // The predicate is mandatory, scheduler, kubelet and other customized scheduler
  // should include this 
  Mandatory() bool
  // Validate whether pod is valid, e.g. selector
  Validate(pod v1.Pod) bool
  Fit(pod v1.Pod, node v1.Node) bool
  // others ??
}
```

Regarding the performance/throughput, there's still some room for us to improve without parallel, e.g. in https://github.com/kubernetes/kubernetes/issues/42002#issuecomment-306106033 , the time is reduced from **10+m** to **~160s** for 50k Pods on 5k nodes (NodeAffinity predicates), I just removed duplicated selector validation of Pod :).

**Comment by misterikkit** [comment]
Hi, it seems like some work was already done to parallelize priority evaluation, but we think we can do better? I know we measure priority evaluation time with the scheduling_algorithm_priority_evaluation metric. Is that tracked in e2e tests?

Finally, I really like the idea for predicates in https://github.com/kubernetes/kubernetes/issues/24246#issuecomment-338635435, but when I brought that up with @bsalamat, he was worried that having ConditionChanged per predicate would cause a lot of duplicated code & performance hit.

**Comment by k82cn** [comment]
> ConditionChanged per predicate would cause a lot of duplicated code & performance hit.

for duplicated codes, it should be fine by some helper funcs. for performance, it should be acceptable :) 

Another option is to have some pre-defined events, e.g. NodeConditionChanged, NodeTaintChanged (defined those event based on what conditions are checked in scheduler), and each predicate just return which event will trigger cache refresh :)

```go
// Event return the event list that trigger cache refresh in this predicate.
func Event() []algorithm.Event {
  return []algorithm.Event{
    algorithm.NodeConditionChanged,
    algorithm.NodeTaintsChanged,
  }
}
```

Scheduler will maintain a map: the key is Event, the value is predicate names. So the overall loop is similar with current logic. But we need to update scheduler to support new event if necessary.

**Comment by k82cn** [comment]
xref https://github.com/kubernetes/kubernetes/issues/62092
